File: skills/webhook-consumer-architecture/examples/go/net-http/idempotency.go
package examples

import (
	"log/slog"
	"net/http"
)

func HandleWebhookRequest(
	w http.ResponseWriter,
	r *http.Request,
	logger *slog.Logger,
	cache WebhookEventCache,
	mapper WebhookEventMapper,
	pipeline WebhookPipeline,
	body map[string]any,
) {
	content, _ := body["content"].(map[string]any)
	eventName, _ := body["event_type"].(string)
	eventID, _ := body["id"].(string)
	eventType := mapper.GetEventType(eventName)

	if eventID == "" || eventType == "" {
		w.WriteHeader(http.StatusOK)
		return
	}

	if cache.Exists(eventID) {
		logger.Info("duplicate webhook event", "event_id", eventID)
		w.WriteHeader(http.StatusOK)
		return
	}

	event := NormalizeSubscriptionEvent(
		ProviderSubscriptionPayload{
			ID:         content["id"].(string),
			CustomerID: content["customer_id"].(string),
			PlanID:     content["plan_id"].(string),
			Status:     content["status"].(string),
		},
		eventID,
		eventType,
	)

	defer cache.Put(eventID)
	_ = pipeline.Process(r.Context(), event)
}

File: skills/webhook-consumer-architecture/examples/python/django/idempotency.py
import logging

from rest_framework.status import HTTP_200_OK

logger = logging.getLogger(__name__)


def handle_webhook_request(request):
    data = request.data
    content = data.get("content")
    event_name = data.get("event_type")
    event_id = data.get("id")
    event_type = webhook_event_mapper.get_event_type(event_name)

    if not event_id or not event_type:
        raise ViewException("Unhandled event", status_code=HTTP_200_OK)

    if webhook_event_cache.get_cache(event_id):
        logger.info("Duplicate webhook event", extra={"event_id": event_id})
        raise ViewException("Duplicated event", status_code=HTTP_200_OK)

    normalized_event = normalize_subscription_event(content, event_id, event_type)

    try:
        webhook_pipeline.process(normalized_event)
    finally:
        webhook_event_cache.set_cache(event_id)
